File: src/main/kotlin/com/nexsabre/hardwarereservationtool/server/api/v1/Machines.kt
package com.nexsabre.hardwarereservationtool.server.api.v1

import com.nexsabre.hardwarereservationtool.server.configuration.Configuration
import com.nexsabre.hardwarereservationtool.server.controllers.ReservationMachine
import com.nexsabre.hardwarereservationtool.server.controllers.allMachines
import com.nexsabre.hardwarereservationtool.server.models.Element
import com.nexsabre.hardwarereservationtool.server.requests.AddMachineRequest
import com.nexsabre.hardwarereservationtool.server.requests.PasswordBodyRequest
import io.swagger.v3.oas.annotations.Operation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@RequestMapping("/api/v1/machines")
class Machines {
    @GetMapping
    @Operation(summary = "Return a list of all devices")
    fun getAllMachines(): List<Element> {
        return allMachines()
    }

    @RequestMapping(method = [RequestMethod.POST], consumes = ["application/json"])
    @Operation(summary = "Create a new machine")
    fun postAddMachine(@RequestBody machine: AddMachineRequest): ResponseEntity<String> {
        val status = ReservationMachine().create(name = machine.name, address = machine.address, null, null)
        return when (status.first) {
            true -> ResponseEntity.created(URI("${status.second.id}")).body(Json.encodeToString<Element>(status.second))
            false -> ResponseEntity(HttpStatus.CONFLICT)
        }
    }

    @GetMapping("/{machineId}")
    @Operation(summary = "Get detailed information about a specific machine")
    fun getSpecificMachine(@PathVariable machineId: Int): ResponseEntity<Element?> {
        val specificMachine = ReservationMachine().get(machineId) ?: return ResponseEntity(HttpStatus.NOT_FOUND)
        return ResponseEntity.ok().body(specificMachine)
    }

    @GetMapping("/{machineId}/delete")
    @Operation(summary = "Delete a machine with specific ID")
    fun deleteMachine(@PathVariable machineId: Int): ResponseEntity<String> {
        return when (ReservationMachine().delete(machineId)) {
            true -> ResponseEntity(HttpStatus.NO_CONTENT)
            false -> ResponseEntity(HttpStatus.OK)
        }
    }

    @GetMapping("/{machineId}/enable")
    @Operation(summary = "Enable machine for reservation operation",
            description = "When user create a new machine, by default it is \'enable\'")
    fun enableMachine(@PathVariable machineId: Int): ResponseEntity<String> {
        return when (ReservationMachine().enable(machineId)) {
            true -> ResponseEntity(HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.CONFLICT)
        }
    }

    @GetMapping("/{machineId}/disable")
    @Operation(summary = "Disable machine for reservation operation",
            description = "Disabled machine can signalize, it is under maintenance")
    fun disableMachine(@PathVariable machineId: Int): ResponseEntity<String> {
        return when (ReservationMachine().disable(machineId)) {
            true -> ResponseEntity(HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.CONFLICT)
        }
    }

    @PostMapping("/{machineId}/protect", consumes = ["application/json"])
    @Operation(summary = "Set a machine in protected state",
            description = "If enabled, the only way to reserve it is using a request with master password")
    fun protectMachine(@PathVariable machineId: Int, @RequestBody passwordRequest: PasswordBodyRequest):
            ResponseEntity<String> {
        if (!Configuration().checkPassword(passwordRequest.password)) {
            return ResponseEntity(HttpStatus.UNAUTHORIZED)
        }
        return when (ReservationMachine().protect(machineId)) {
            true -> ResponseEntity(HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.CONFLICT)
        }
    }

    @PostMapping("/{machineId}/unprotect", consumes = ["application/json"])
    @Operation(summary = "Release a machine from the protected state",
            description = "No password is needed to reserve a machine")
    fun unprotectMachine(@PathVariable machineId: Int, @RequestBody passwordRequest: PasswordBodyRequest):
            ResponseEntity<String> {
        if (!Configuration().checkPassword(passwordRequest.password)) {
            return ResponseEntity(HttpStatus.UNAUTHORIZED)
        }
        return when (ReservationMachine().unprotect(machineId)) {
            true -> ResponseEntity(HttpStatus.OK)
            false -> ResponseEntity(HttpStatus.CONFLICT)
        }
    }
}

File: e2e/api_cmds/machines.py
import json
import requests

from api_cmds import configuration as c


def _change_protected(machine_id: int, password, state: bool) -> bool:
    state = "protect" if state else "unprotect"
    r = requests.post(c.api_v1(f"/machines/{machine_id}/{state}"), json={"password": password})
    return True if r.status_code == 200 else False


def protect(machine_id: int, password: str) -> bool:
    return _change_protected(machine_id, password, True)


def unprotect(machine_id: int, password: str) -> bool:
    return _change_protected(machine_id, password, False)


def _change_enabled(machine_id, status):
    status = "enable" if status else "disable"
    r = requests.get(c.api_v1(f"/machines/{machine_id}/{status}"))
    return True if r.status_code == 200 else False


def enable(machine_id):
    return _change_enabled(machine_id, True)


def disable(machine_id):
    return _change_enabled(machine_id, False)


def get_machine(machine_id):
    r = requests.get(c.api_v1(f"/machines/{machine_id}"))
    try:
        return json.loads(r.content)
    except Exception as e:
        print(e)
        return {}


def get_all_machines():
    content = requests.get(c.api_v1("/machines")).content
    return json.loads(content)


def add_example_machine():
    r = requests.post(c.api_v1("/machines"), json={
        "name": c.random_name(),
        "address": c.random_ip()
    })
    assert r.status_code == 201, "Return code should be Created"
    try:
        return json.loads(r.content)
    except Exception as e:
        print(e)
        return {}


def add_and_verifies_new_machine():
    l_before = count_machines()
    example_machine = add_example_machine()
    l_after = count_machines()
    assert l_after > l_before, "One machine should be added"
    return example_machine


def count_machines() -> int:
    return len(get_all_machines())


def remove_specific_machine(machine_id):
    r = requests.get(c.api_v1(f"/machines/{machine_id}/delete"))
    return True if r.status_code == 204 else False


def remove_all_machines():
    all_machines = get_all_machines()
    for machine in all_machines:
        r = requests.get(c.api_v1(f"/machines/{machine['id']}/delete"))
        assert r.status_code == 204, "Should return No Content"

    assert count_machines() == 0, "All machines should be removed"
